src/types/routerPush.ts

```diff
@@ -8,9 +9,8 @@ export type RouterPushOptions = {
 
 export type RouterPush<
   TRoutes extends RouterRoutes
-> = <
-  TRoutePath extends string
->(source: string | RouteWithParams<TRoutes, TRoutePath>, options?: RouterPushOptions) => Promise<void>
-
-// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-arguments
-export type RegisteredRouterPush = RouterPush<RegisteredRoutes>
\ No newline at end of file
+> = {
+  <TSource extends RouteKeysThatHaveOptionalParams<TRoutes>>(source: TSource, params?: RouteParamsByName<TRoutes, TSource>, options?: RouterPushOptions): Promise<void>,
+  <TSource extends RouteKeysThatHaveRequireParams<TRoutes>>(source: TSource, params: RouteParamsByName<TRoutes, TSource>, options?: RouterPushOptions): Promise<void>,
```

Comment:
Does this still show all the route keys in the type hint even though they are separated into two different overloads?

Reply:
yeah it works pretty well, I'll see if there's a way that provides an even better syntax

https://github.com/kitbagjs/router/assets/6098901/dcfaca6a-b3c0-4833-a48c-e7a41c08d406

Reply:
this type seems to be a bit better dx?

https://github.com/kitbagjs/router/assets/6098901/e9f4d0c9-7844-49b4-99f6-81969802a8e9

<img width="736" alt="image" src="https://github.com/kitbagjs/router/assets/6098901/613bd05e-24dd-483f-ba3f-7670fc7ec630">

Reply:
yeah this looks great